src/main/scala/com/databricks/sparkdl/ModelFetcher.scala

```diff
@@ -0,0 +1,112 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.databricks.sparkdl
+
+import java.io.{File, FileOutputStream}
+import java.net.URL
+import java.nio.channels.Channels
+import java.nio.file.{Files, Path, Paths}
+import java.security.{DigestInputStream, MessageDigest}
+import java.util.Base64
+
+import org.apache.commons.io.IOUtils
+import org.tensorflow.framework.GraphDef
+import org.tensorframes.impl.{SerializedGraph, TensorFlowOps}
+
+object ModelFetcher {
+
+  private val cacheDirectory = ".spark-deep-learning"
+
+  private[sparkdl] class ModelFetchException(message: String) extends Exception(message)
+
+  /**
+   * Returns the absolute path of a directory in the user's home directory where we can download
+   * models. If the directory does not already exist, create the directory.
+   */
+  private[sparkdl] def getOrCreateModelCacheDir: String = {
+    val home = System.getProperty("user.home")
+    val modelDir = new File(home, cacheDirectory)
+    modelDir.mkdirs()
+    modelDir.getAbsolutePath()
+  }
+
+  /**
+   * Import GraphDef from file and check the checksum matches.
+   *
+   * @param filePath Path of graph file.
+   * @param base64Hash Checksum to validate file.
+   * @return GraphDef if checksum matches or None.
+   */
+  private[sparkdl] def importGraph(filePath: Path, base64Hash: String): Option[GraphDef] = {
+    var bytes: Array[Byte] = null
+    val md = MessageDigest.getInstance("SHA-256")
+    val is = Files.newInputStream(filePath)
+    val dis = new DigestInputStream(is, md)
+    try {
+      bytes = IOUtils.toByteArray(dis)
+    } finally {
+      is.close()
+      dis.close()
+    }
+    val digest = Base64.getEncoder.encodeToString(md.digest)
+    if (digest == base64Hash) {
+      Some(TensorFlowOps.readGraphSerial(SerializedGraph.create(bytes)))
+    } else {
+      None
+    }
+  }
+
+  /**
+   * Get GraphDef from URL. The file is already in the cache directory and the checksum matches,
+   * we'll skip the download. Otherwise the file will be downloaded to the cache directory. If
+   * the checksum does not match after download, raise a ModelFetchException.
+   *
+   * @param source url source for model.
+   * @param fileName file path where we will look for or cache the model file.
+   * @param base64Hash check sum to verify file contents.
+   * @return model as a GraphDef.
+   */
+  private[sparkdl] def getFromWeb(
+    source: String,
+    fileName: String,
+    base64Hash: String): GraphDef = {
+
+    val filePath = Paths.get(getOrCreateModelCacheDir, fileName)
+    val graph = if (Files.exists(filePath)) {
+      importGraph(filePath, base64Hash)
+    } else {
+      None
+    }
+    graph.getOrElse {
+      // Graph file did not exist or had the wrong hash, try downloading from source.
+      val url = new URL(source)
+      val urlChannel = Channels.newChannel(url.openStream)
```

Comment:
Do we want the ability to set a timeout here/should users be able to specify a timeout for the download? I'm not sure exactly how Java sets its default timeouts & it might matter for larger models.

Potentially useful: from the [Java docs](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#openStream()), `url.openStream` is a convenience for calling `url.getConnection().getInputStream()`; `url.getConnection()` returns a [`URLConnection`](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html) upon which we can call `setTimeout()` to set a timeout.